File: backend/models/device_grant.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// DeviceGrant is a long-lived, revocable per-install credential (issue #722's
// "fully biometric login"): minted for an authenticated device so that a
// successful local biometric unlock can be exchanged — via
// POST /auth/device/session — for a fresh session JWT past the stored token's
// expiry. Follows ApiToken exactly: only the SHA-256 hash is stored, the row
// is soft-deleted (gorm.Model), and grants are revoked with RevokedAt rather
// than removed, so revocation is auditable. The plaintext token is returned
// exactly once at creation and never persisted or logged.
type DeviceGrant struct {
	gorm.Model
	UserID     uint   `gorm:"not null"`
	TokenHash  string `gorm:"not null;unique" json:"-"`
	Label      string `gorm:"not null;default:''"`
	LastUsedAt *time.Time
	RevokedAt  *time.Time
}

File: android/core/model/src/main/kotlin/com/mycorrhizal/crm/model/network/DeviceGrant.kt
package com.mycorrhizal.crm.model.network

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

// --- Issue #722 fully-biometric-login device grants (server half) ---

/** POST /api/v1/auth/device/grants body — optional display label. */
@JsonClass(generateAdapter = true)
data class DeviceGrantCreateRequest(
    val label: String? = null,
)

/**
 * POST /api/v1/auth/device/grants response. [token] is the one-time plaintext
 * grant, returned exactly once; [id] lets the client revoke this device later.
 */
@JsonClass(generateAdapter = true)
data class DeviceGrantCreateResponse(
    val id: Long = 0,
    val label: String? = null,
    @Json(name = "created_at") val createdAt: String? = null,
    @Json(name = "last_used_at") val lastUsedAt: String? = null,
    @Json(name = "revoked_at") val revokedAt: String? = null,
    val token: String? = null,
)

/** POST /api/v1/auth/device/session body — possession of the stored grant. */
@JsonClass(generateAdapter = true)
data class DeviceGrantExchangeRequest(
    @Json(name = "device_token") val deviceToken: String,
)

/** POST /api/v1/auth/device/grants/revoke-all — `{ revoked: n }`. */
@JsonClass(generateAdapter = true)
data class RevokeAllDeviceGrantsResponse(
    val revoked: Int? = null,
)
